### ISSUE TITLE
Squat - audio feedback on angles

### ISSUE THREAD

**Issue by csabau** [csabau/BTracking#8]
I've created a new file based on the working one called `ARSUIViewSquat`. I'll be keeping `ARSUIView2D` as a foundation / safety file for now. 

In the Squat file I've displayed angles for all the joints I'm looking to analyse and renamed the function [makeJointAngleVisible](https://github.com/csabau/BTracking/blob/a02e0b2fa29578089da3d66aa27a68320e8d6e19/BodyTracking-Example/BodyTracking-Example/ARSUIViewSquat.swift#L54-L87). 

I'm also displaying only the bones from the right side of the body and renamed the bones dictionary to [squatBonesToShow](https://github.com/csabau/BTracking/blob/a02e0b2fa29578089da3d66aa27a68320e8d6e19/BodyTracking-Example/BodyTracking-Example/ARSUIViewSquat.swift#L158-L184).

Now I'm not entirely sure what the logic / maths for a perfect squat is just yet - I have an idea, but I need to work out the actual maths behind it. It will be based on angles between joints, and I have the value for those, so in broad terms I will be doing some sort of comparison. 

I am currently thinking to add some audio feedback when and angle is wrong / outside the margin of error. 
At the moment I'm thinking of 2 potentially 3 angles give feedback on:
- hip angle
- knee angle 
- shoulder - foot / knee relationship
This will imply 3 different sounds. Can you suggest resources on how to make my app sing?
Any other thoughts on tracking from / giving feedback?

I'm thinking of having minimal sound based feedback during the exercise , and to somehow keep count of how many time a specific angle goes out of the margin on error, to provide more written feedback afterwards...potentially

**Comment by mhamilt** [comment]
> I am currently thinking to add some audio feedback

You can make something a bit more bespoke down the line, but for the time being you could just `import AVFoundation` and use `AudioServicesPlaySystemSound(1111)` whenever you want a sound to play

**Comment by csabau** [comment]
Thank you , that was really helpful. I've managed to define a range for a good neck posture, and whenever the neck is out of range, the sound plays in the [checkForm function](https://github.com/csabau/BTracking/blob/47dea508cf8c8318603aa2aacd00c2a3fee6d798/BodyTracking-Example/BodyTracking-Example/ARSUIViewSquat.swift#L87-L90) this is now commented out because I have a call of the checkForm() in the [Delegate function](https://github.com/csabau/BTracking/blob/47dea508cf8c8318603aa2aacd00c2a3fee6d798/BodyTracking-Example/BodyTracking-Example/ARSUIViewSquat.swift#L238). The problem with that is that the sound plays every frame (if the neck is out of range) so the experience is not too great. I would need to check whether the neck is in position every frame, but can I play the sound ONCE, and not have it play again for 3s or so even if it's still out of range?

**Comment by csabau** [comment]
The rest of the [checkForm function](https://github.com/csabau/BTracking/blob/47dea508cf8c8318603aa2aacd00c2a3fee6d798/BodyTracking-Example/BodyTracking-Example/ARSUIViewSquat.swift#L84-L99) is now essentially using a new function I've created in the `BodyTracker2D` which calculates the distance between 2 CGPoints [distanceBetween2Joints](https://github.com/csabau/BTracking/blob/47dea508cf8c8318603aa2aacd00c2a3fee6d798/Sources/BodyTracking/BodyTracker2D.swift#L265-L281). What I'm intending to do with this is have the user being "scanned" before they perform the exercise, in order to work out body proportions between the lower part of the leg, the upper part and the torso. 

These are really important as they define what rapport / ratio the hip and knee angles take. For my personal proportions (which are considered quite even) from testing it looks like the reading of the angles at the hip and knee are identical. Therefore for a person with similar proportions, their angles should be within maybe 5 degrees of each other for a proper form which will indicate that the hip hinge is working in tandem with the bending of the knee. 

So say if the knee angle goes below the 5 degrees difference according to the hip angle, that would mean the torso may be too upright and will add pressure to the knee so feedback would be - bend from the hips more. 
If the hip angle would go below 5 degrees, they're bending too far forward so feedback would be - bend at the knee more, lower the hips.

However, different limb ratios would imply different angles and that's when things can get really tricky. I'm trying to get some mates with different proportions to help me do some reading to see what the correlation is , what sort of ratios would imply a significant difference on angles. But I can easily lose a lot of time on that....

Currently the reading is being called in the init with a delay of 5s, so I print out just one set of numbers. 
I've done 6 readings and the ratios are quite consistent 

```
Up leg legth: 226.83131387158932  - 3
Down leg legth: 194.7546278907481 - 3.49
Torso legth: 258.9618590824301 - 2.62
Total: 680.54780084476752 (678)

Up leg legth: 265.23016090508327 - 2.7
Down leg legth: 175.26388252355673 - 4
Torso legth: 274.1741862129957 - 2.6
Total: 714

Up leg legth: 236.3690355229205 - 2.95
Down leg legth: 194.49133977062672 - 3.59
Torso legth: 268.5738434688755 - 2.6
Total: 698

Up leg legth: 206.28706689364276 - 3.08
Down leg legth: 192.69951299335028 - 3.3
Torso legth: 238.97757078516275 - 2.6
Total: 636



Up leg legth: 219.52803782739412 - 2.94
Down leg legth: 182.9559965107557 -3.5
Torso legth: 244.60594317871556 -2.6
Total 645
```

So I'm inclined to code something along the lines of if the ratios are between 2.4 and 3.6 - make the angles needing to be identical, with a 5 degree margin of error, maybe 10 degrees.


Any thoughts?

**Comment by mhamilt** [comment]
> I would need to check whether the neck is in position every frame, but can I play the sound ONCE, and not have it play again for 3s or so even if it's still out of range?

Definitely, just some kind of `if` guard will do. Create a boolean and change it similarly to how you have here

https://github.com/csabau/BTracking/blob/47dea508cf8c8318603aa2aacd00c2a3fee6d798/Sources/BodyTracking/BodyTracker2D.swift#L137-L141

But you'd probably want to have two booleans, one for if the neck is in frame `neckIsInFrame` and another to tell you if the neck was previously in the frame `neckWasPreviouslyInFrame`

so something like

```swift

if(!neckIsInFrame)
{
   if(neckWasPreviouslyInFrame)
   {
    playNeckOutOfFrameSound()
    neckWasPreviouslyInFrame = false;
   }
}
else
{
   if(!neckWasPreviouslyInFrame)
   {
    playNeckIsInFrameSound()
    neckWasPreviouslyInFrame = true;
   }
}
```

**Comment by mhamilt** [comment]
>Currently the reading is being called in the init with a delay of 5s, so I print out just one set of numbers.

Are these numbers length-in-pixels or something else?

**Comment by csabau** [comment]
> > Currently the reading is being called in the init with a delay of 5s, so I print out just one set of numbers.
> 
> Are these numbers length-in-pixels or something else?

These are just distances between CGPoints, so I would assume length in pixels. 

I only need them to determine body ratios anyway, and I think I'm getting somewhere to make the reading more customisable to the user. 
Currently I'm dealing with even proportions case, because I get the readings of my own body. So when the proportions between the limbs are within some parameters such as lower leg 20-35% , upper leg 25-40% and torso 30-45% those include some margin of errors from the reading that would still indicate that the bombs are roughly 30% each, with the lower limbs being slightly shorter and torso slightly longer - that's when I want to run the function that makes sure that the hip and knee angle are within 10 degrees of each other at all times.
But to implement that and see how it works, I need the function to be run every frame to constantly check the angles, and make a sound when the difference is greater than 10, but only once, and if the issue consists, play it again say 2s later
